File: Sources/custom_adapter/TopOnAdapter/Objc/6.5.x/Interstitial/AlxTopOnInterstitialAdapter.m
//
//  AlxTopOnInterstitialAdapter.m
//  AlxAdsOCDemo
//

#import "AlxTopOnInterstitialAdapter.h"
#import "AlxTopOnBaseManager.h"
#import "AlxTopOnInterstitialDelegate.h"
#import "AlxTopOnInterstitialEvent.h"
#import "AlxTopOnBiddingRequest.h"
#import "AlxTopOnBiddingRequestManager.h"
#import "AlxTopOnTool.h"

static NSString *const TAG = @"AlxTopOnInterstitialAdapter";

@interface AlxTopOnInterstitialAdapter ()
// Alx SDK 插屏广告对象 / Alx SDK interstitial ad object
@property (nonatomic, strong) AlxInterstitialAd *interstitialAd;
// 广告事件代理 / Ad event delegate
@property (nonatomic, strong) AlxTopOnInterstitialDelegate *interstitialDelegate;
@end

@implementation AlxTopOnInterstitialAdapter

#pragma mark - 懒加载 / Lazy Load

- (AlxTopOnInterstitialDelegate *)interstitialDelegate {
    if (_interstitialDelegate == nil) {
        _interstitialDelegate = [[AlxTopOnInterstitialDelegate alloc] init];
        _interstitialDelegate.adStatusBridge = self.adStatusBridge;
    }
    return _interstitialDelegate;
}

#pragma mark - 广告加载 / Ad Load

- (void)loadADWithArgument:(ATAdMediationArgument *)argument {
    NSLog(@"%@: loadADWithArgument", TAG);
    NSLog(@"%@: loadAD: isMainThread=%@", TAG, [NSThread isMainThread] ? @"YES" : @"NO");
    
    NSString *bidId = argument.serverContentDic[kATAdapterCustomInfoBuyeruIdKey];
    NSLog(@"%@: loadAD: bidId=%@", TAG, bidId ?: @"空");
    
    dispatch_async(dispatch_get_main_queue(), ^{
        NSString *unitKeyStr = [AlxTopOnBaseManager unitID];
        NSString *unitId = argument.serverContentDic[unitKeyStr];
        
        if (!unitId || unitId.length == 0) {
            NSString *errorStr = @"unitid is empty";
            NSLog(@"%@: loadAD: error = %@", TAG, errorStr);
            NSError *error = [AlxTopOnBaseManager errorWithCode:-100 message:errorStr];
            [self.adStatusBridge atOnAdLoadFailed:error adExtra:@{}];
            return;
        }
        NSLog(@"%@: loadAD: unitid = %@", TAG, unitId);
        
        if (bidId) {
            // Bidding 场景：从缓存中取出已加载的广告
            // Bidding scenario: retrieve the pre-loaded ad from cache
            AlxTopOnBiddingRequest *biddingRequest = [[AlxTopOnTool shared] getRequestItemWithUnitID:unitId];
            if (biddingRequest) {
                self.interstitialAd = (AlxInterstitialAd *)biddingRequest.customObject;
                
                if (self.interstitialAd) {
                    NSLog(@"%@: loadAD: bid ad loaded, notify success", TAG);
                    // ⚠️ 注意：将广告对象传给 TopOn SDK
                    // ⚠️ Note: Pass the ad object to the TopOn SDK
                    NSMutableDictionary *adExtra = [NSMutableDictionary dictionary];
                    adExtra[kATAdAssetsCustomObjectKey] = self.interstitialAd;
                    [self.adStatusBridge atOnInterstitialAdLoadedExtra:adExtra];
                } else {
                    NSLog(@"%@: loadAD: bid ad object is empty", TAG);
                    NSError *error = [NSError errorWithDomain:@"AlxTopOnAdapter" code:-100 userInfo:@{NSLocalizedDescriptionKey: @"Bid ad object is empty"}];
                    [self.adStatusBridge atOnAdLoadFailed:error adExtra:@{}];
                }
            } else {
                NSLog(@"%@: loadAD: bid request not found in cache", TAG);
                NSError *error = [NSError errorWithDomain:@"AlxTopOnAdapter" code:-100 userInfo:@{NSLocalizedDescriptionKey: @"Bid request not found"}];
                [self.adStatusBridge atOnAdLoadFailed:error adExtra:@{}];
            }
            [[AlxTopOnTool shared] removeRequestItemWithUnitID:unitId];
        } else {
            // 普通加载场景 / Normal (non-bidding) load scenario
            self.interstitialAd = [[AlxInterstitialAd alloc] init];
            self.interstitialAd.delegate = self.interstitialDelegate;
            // 设置 Delegate 的广告对象引用 / Set the ad object reference on the delegate
            self.interstitialDelegate.interstitialAd = self.interstitialAd;
            
            NSLog(@"%@: start loading ad with unitId: %@", TAG, unitId);
            [self.interstitialAd loadAdWithAdUnitId:unitId];
        }
    });
}

#pragma mark - C2S 竞价 / C2S Bidding

+ (void)bidRequestWithPlacementModel:(ATPlacementModel *)placementModel
                      unitGroupModel:(ATUnitGroupModel *)unitGroupModel
                                info:(NSDictionary *)info
                          completion:(void (^)(ATBidInfo * _Nullable, NSError * _Nullable))completion {
    NSLog(@"%@: bidRequestWithPlacementModel", TAG);
    
    dispatch_async(dispatch_get_main_queue(), ^{
        if (![AlxTopOnBaseManager isInitialized]) {
            [AlxTopOnBaseManager initSDKWithServerInfo:info];
        }
    });
    
    AlxTopOnInterstitialEvent *customEvent = [[AlxTopOnInterstitialEvent alloc] initWithInfo:info localInfo:info];
    customEvent.isC2SBiding = YES;
    
    AlxTopOnBiddingRequest *request = [[AlxTopOnBiddingRequest alloc]
                                       initWithUnitGroup:unitGroupModel
                                       customEvent:customEvent
                                       unitID:info[[AlxTopOnBaseManager unitID]]
                                       placementID:placementModel.placementID
                                       extraInfo:info
                                       adType:ATAdFormatInterstitial
                                       bidCompletion:completion];
    
    [[AlxTopOnBiddingRequestManager shared] startWithRequest:request];
}

#pragma mark - 广告就绪检查（实例方法）/ Ad Ready Check (Instance Method)

- (BOOL)adReadyInterstitialWithInfo:(NSDictionary *)info {
    NSLog(@"%@: adReadyInterstitialWithInfo", TAG);
    
    // 检查广告对象是否准备好 / Check whether the ad object is ready
    if (self.interstitialAd) {
        NSLog(@"%@: adReady = YES", TAG);
        return YES;
    }
    
    NSLog(@"%@: adReady = NO", TAG);
    return NO;
}

#pragma mark - 展示广告（实例方法）/ Show Ad (Instance Method)

- (void)showInterstitialInViewController:(UIViewController *)viewController {
    NSLog(@"%@: showInterstitialInViewController", TAG);
    
    dispatch_async(dispatch_get_main_queue(), ^{
        if (self.interstitialAd) {
            // 展示广告 / Show the ad
            [self.interstitialAd showAdWithPresent:viewController];
            NSLog(@"%@: ad shown", TAG);
        } else {
            NSLog(@"%@: interstitialAd is nil", TAG);
        }
    });
}

@end

File: custom_adapter/TopOnAdapter/Swift/6.4.x/AlxTopOnInterstitialAdapter.swift
//
//  AlxTopOnInterstitialAdapter.swift
//

import Foundation
import AnyThinkSDK
import AnyThinkInterstitial
import AlxAds

@objc(AlxTopOnInterstitialAdapter)
public class AlxTopOnInterstitialAdapter: ATAdAdapter {
    
    private static let TAG = "AlxTopOnInterstitialAdapter"
    
    private var interstitialAd: AlxInterstitialAd? = nil
    private var customEvent:AlxTopOnInterstitialEvent? = nil
    
    
    // MARK: - Initialization
    @objc public required init(networkCustomInfo serverInfo: [AnyHashable: Any],
                        localInfo: [AnyHashable: Any]) {
        super.init()
        NSLog("%@: init",AlxTopOnInterstitialAdapter.TAG)
        NSLog("%@: init: isMainThread=%@",AlxTopOnInterstitialAdapter.TAG,Thread.current.isMainThread ? "YES" : "NO")
        DispatchQueue.main.async {
            if !AlxTopOnBaseManager.isInitialized {
                AlxTopOnBaseManager.initSDK(serverInfo: serverInfo)
            }
        }
    }
    
    /// 加载广告
    @objc public func loadAD(withInfo serverInfo: [AnyHashable: Any],
                             localInfo: [AnyHashable: Any],
                             completion: @escaping ([[AnyHashable: Any]]?, (any Error)?) -> Void) {
        NSLog("%@: loadAD",AlxTopOnInterstitialAdapter.TAG)
        NSLog("%@: loadAD: isMainThread=%@",AlxTopOnInterstitialAdapter.TAG,Thread.current.isMainThread ? "YES" : "NO")
       
        DispatchQueue.main.async {
            if !AlxTopOnBaseManager.isInitialized {
                AlxTopOnBaseManager.initSDK(serverInfo: serverInfo)
            }
            
            let bidId=serverInfo[kATAdapterCustomInfoBuyeruIdKey] as? String
            NSLog("%@: loadAD: bidId=%@",AlxTopOnInterstitialAdapter.TAG,bidId ?? "")
            
            guard let unitId = serverInfo[AlxTopOnBaseManager.unitID] as? String,!unitId.isEmpty else{
                let errorStr="unitid is empty"
                NSLog("%@: loadAD: error= %@",AlxTopOnInterstitialAdapter.TAG,errorStr)
                completion(nil,AlxTopOnBaseManager.error(code: -100, msg: errorStr))
                return
            }
            NSLog("%@: loadAD: unitid=%@",AlxTopOnInterstitialAdapter.TAG,unitId)
            
            if bidId != nil {
                let request:AlxTopOnBiddingRequest? = AlxTopOnTool.shared.getRequestItem(withUnitID: unitId) as? AlxTopOnBiddingRequest
                if let request = request {
                    self.customEvent = request.customEvent as? AlxTopOnInterstitialEvent
                    self.customEvent?.requestCompletionBlock = completion
                    self.interstitialAd = request.customObject as? AlxInterstitialAd
                    //判断广告源是否已经loaded过
                    if let interstitialAd = self.interstitialAd {
                        self.customEvent?.trackInterstitialAdLoaded(interstitialAd, adExtra: nil)
                    }else{
                        NSLog("%@: loadAD: bid ad object is empty",AlxTopOnInterstitialAdapter.TAG)
                    }
                }
                AlxTopOnTool.shared.removeRequestItem(withUnitID: unitId)
            }else{
                self.customEvent = AlxTopOnInterstitialEvent(info: serverInfo, localInfo: localInfo)
                self.customEvent?.requestCompletionBlock = completion
                
                // load ad
                self.interstitialAd = AlxInterstitialAd()
                self.interstitialAd?.delegate = self.customEvent
                self.interstitialAd?.loadAd(adUnitId: unitId)
            }
        }
    }
    
    // MARK: - C2S header bidding 竞价
    @objc public static func bidRequestWithPlacementModel(_ placementModel: ATPlacementModel,unitGroupModel: ATUnitGroupModel, info: [AnyHashable: Any], completion: @escaping (ATBidInfo?, Error?) -> Void) {
        NSLog("%@: bidRequestWithPlacementModel",AlxTopOnInterstitialAdapter.TAG)
        NSLog("%@: bidRequestWithPlacementModel: isMainThread=%@",AlxTopOnInterstitialAdapter.TAG,Thread.current.isMainThread ? "YES" : "NO")
        
        DispatchQueue.main.async {
            if !AlxTopOnBaseManager.isInitialized {
                AlxTopOnBaseManager.initSDK(serverInfo: info)
            }
        }
        
        let customEvent:AlxTopOnInterstitialEvent = AlxTopOnInterstitialEvent(info: info, localInfo: info)
        customEvent.isC2SBiding = true
        let request:AlxTopOnBiddingRequest = AlxTopOnBiddingRequest(
            unitGroup: unitGroupModel,
            customEvent: customEvent,
            unitID: info[AlxTopOnBaseManager.unitID] as? String,
            placementID: placementModel.placementID,
            extraInfo: info,
            adType: ATAdFormat.interstitial,
            bidCompletion: completion
        )
        AlxTopOnBiddingRequestManager.shared.start(with: request)
    }
    
    @objc(loadWithServerInfo:localInfo:)
    public func loadWithServerInfo(_ serverInfo:[AnyHashable: Any],localInfo:[AnyHashable: Any]){
        NSLog("%@: loadWithServerInfo",AlxTopOnInterstitialAdapter.TAG)
    }
    
    
    // MARK: - 广告就绪检查
    @objc public static func adReady(withCustomObject customObject: Any, info: [AnyHashable: Any]) -> Bool {
        // 检查广告是否就绪
        // 实际实现中应根据广告网络的具体方法进行检查
        NSLog("%@: adReady",AlxTopOnInterstitialAdapter.TAG)
        if customObject as? AlxInterstitialAd != nil {
            NSLog("%@: adReady true",AlxTopOnInterstitialAdapter.TAG)
            return true
        }else{
            NSLog("%@: adReady false",AlxTopOnInterstitialAdapter.TAG)
            return false
        }
    }
    
    // MARK: - 广告展示
    @objc(showInterstitial:inViewController:delegate:)
    public static func showInterstitial(_ interstitial: ATInterstitial,
                                       in viewController: UIViewController,
                                       delegate: ATInterstitialDelegate) {
        NSLog("%@: showInterstitial",AlxTopOnInterstitialAdapter.TAG)
        guard let interstitialAd = (interstitial.customObject as? AlxInterstitialAd) else {
            NSLog("%@: showInterstitial: interstitialAd object is nil",AlxTopOnInterstitialAdapter.TAG)
            return
        }
        interstitial.customEvent.delegate = delegate
        if interstitialAd.isReady() {
            interstitialAd.showAd(present: viewController)
        }
    }
}
